Compilation errors on prerelease PowerPC Mac OS X Snow Leopard

Compilation fails on this:
```
In file included from ass_utils.c:30:
ass.h:421: error: wrong number of arguments specified for ‘deprecated’ attribute
ass.h:520: error: wrong number of arguments specified for ‘deprecated’ attribute
In file included from ass.c:35:
ass.h:421: error: wrong number of arguments specified for ‘deprecated’ attribute
ass.h:520: error: wrong number of arguments specified for ‘deprecated’ attribute
libtool: compile:  /usr/bin/gcc-4.2 -std=gnu99 -DHAVE_CONFIG_H -I. -I.. -I/opt/local/include -std=gnu99 -Wall -Wextra -Wno-sign-compare -Wno-unused-parameter -Werror-implicit-function-declaration -Wstrict-prototypes -Wpointer-arith -Wredundant-decls -Wno-missing-field-initializers -D_GNU_SOURCE -pipe -Os -arch ppc -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -I/opt/local/include/fribidi -I/opt/local/include/harfbuzz -I/opt/local/include/glib-2.0 -I/opt/local/lib/glib-2.0/include -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -I/opt/local/include/ossp -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -MT ass_string.lo -MD -MP -MF .deps/ass_string.Tpo -c ass_string.c -o ass_string.o >/dev/null 2>&1
make[2]: *** [ass_utils.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
libtool: compile:  /usr/bin/gcc-4.2 -std=gnu99 -DHAVE_CONFIG_H -I. -I.. -I/opt/local/include -std=gnu99 -Wall -Wextra -Wno-sign-compare -Wno-unused-parameter -Werror-implicit-function-declaration -Wstrict-prototypes -Wpointer-arith -Wredundant-decls -Wno-missing-field-initializers -D_GNU_SOURCE -pipe -Os -arch ppc -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -I/opt/local/include/fribidi -I/opt/local/include/harfbuzz -I/opt/local/include/glib-2.0 -I/opt/local/lib/glib-2.0/include -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -I/opt/local/include/ossp -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -MT ass_strtod.lo -MD -MP -MF .deps/ass_strtod.Tpo -c ass_strtod.c -o ass_strtod.o >/dev/null 2>&1
mv -f .deps/ass_string.Tpo .deps/ass_string.Plo
make[2]: *** [ass.lo] Error 1
```

The older version builds:
```
36-47% port -v installed libass
The following ports are currently installed:
  libass @0.13.7_0 (active) requested_variants='' platform='darwin 10' archs='ppc' date='2022-02-14T20:14:23+0800'
```

If using gcc7, build fails a bit further on this:
```
ass_coretext.c: In function 'is_postscript_font_format':
ass_coretext.c:69:25: error: 'kCTFontFormatOpenTypePostScript' undeclared (first use in this function); did you mean 'kCTFontFormatPostScript'?
         ret = format == kCTFontFormatOpenTypePostScript ||
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                         kCTFontFormatPostScript
ass_coretext.c:69:25: note: each undeclared identifier is reported only once for each function it appears in
make[2]: *** [ass_coretext.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
mv -f .deps/ass_render.Tpo .deps/ass_render.Plo
mv -f .deps/ass_blur.Tpo .deps/ass_blur.Plo
```

@TheOneric I have added the patch above, removing my patch, and tried to build with `gcc-4.2`: no, it still fails (on 10.6 PPC):
```
ass_coretext.c: In function ‘is_postscript_font_format’:
ass_coretext.c:69: error: ‘kCTFontFormatOpenTypePostScript’ undeclared (first use in this function)
ass_coretext.c:69: error: (Each undeclared identifier is reported only once
ass_coretext.c:69: error: for each function it appears in.)
make[2]: *** [ass_coretext.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
mv -f .deps/ass_blur.Tpo .deps/ass_blur.Plo
make[2]: Leaving directory `/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_libass/libass/work/libass-libass-98f5f3e/libass'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_libass/libass/work/libass-libass-98f5f3e'
make: *** [all] Error 2
make: Leaving directory `/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_libass/libass/work/libass-libass-98f5f3e'
Command failed:  cd "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_multimedia_libass/libass/work/libass-0.15.2" && /usr/bin/make -j4 -w all 
Exit code: 2
Error: Failed to build libass: command execution failed
Error: See /opt/local/var/macports/logs/_opt_PPCSnowLeopardPorts_multimedia_libass/libass/main.log for details.
Error: Follow https://guide.macports.org/#project.tickets if you believe there
is a bug.
Error: Processing of port libass failed
```

To be precise, this was the only patch used:
```
--- libass/ass.h.orig	2021-09-12 05:34:52.000000000 +0800
+++ libass/ass.h	2022-02-19 23:14:23.000000000 +0800
@@ -31,7 +31,11 @@
 #endif
 
 #if (defined(__GNUC__) && ((__GNUC__ > 3) || (__GNUC__ == 3 && __GNUC_MINOR__ >= 1))) || defined(__clang__)
-    #define ASS_DEPRECATED(msg) __attribute__((deprecated(msg)))
+    #if __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 5) || defined(__clang__)
+        #define ASS_DEPRECATED(msg) __attribute__((deprecated(msg)))
+    #else
+        #define ASS_DEPRECATED(msg) __attribute__((deprecated))
+    #endif
     #if __GNUC__ > 5 || defined(__clang__)
         #define ASS_DEPRECATED_ENUM(msg) __attribute__((deprecated(msg)))
     #else
```

--- a/configure.ac
+++ b/configure.ac
@@ -146,7 +146,7 @@ AS_IF([test "x$enable_coretext" != xno], [
     AC_COMPILE_IFELSE([
         AC_LANG_PROGRAM( dnl# First test for legacy include
             [[#include <ApplicationServices/ApplicationServices.h>]],
-            [[CTFontDescriptorCopyAttribute(NULL, kCTFontURLAttribute);]]
+            [[CTFontDescriptorCopyAttribute(NULL, kCTFontNameAttribute);]]
         )
     ], [
         pkg_libs="$pkg_libs -framework ApplicationServices -framework CoreFoundation"
@@ -158,7 +158,7 @@ AS_IF([test "x$enable_coretext" != xno], [
         AC_COMPILE_IFELSE([
             AC_LANG_PROGRAM( dnl# Otherwise check newer include style
                 [[#include <CoreText/CoreText.h>]],
-                [[CTFontDescriptorCopyAttribute(NULL, kCTFontURLAttribute);]]
+                [[CTFontDescriptorCopyAttribute(NULL, kCTFontNameAttribute);]]
             )
         ], [
             pkg_libs="$pkg_libs -framework CoreText -framework CoreFoundation"

--- a/libass/ass_coretext.c
+++ b/libass/ass_coretext.c
@@ -61,29 +61,6 @@ static void destroy_font(void *priv)
     CFRelease(fontd);
 }
 
-static bool is_postscript_font_format(CFNumberRef cfformat)
-{
-    bool ret = false;
-    int format;
-    if (CFNumberGetValue(cfformat, kCFNumberIntType, &format)) {
-        ret = format == kCTFontFormatOpenTypePostScript ||
-              format == kCTFontFormatPostScript;
-    }
-    return ret;
-}
-
-static bool check_postscript(void *priv)
-{
-    CTFontDescriptorRef fontd = priv;
-    CFNumberRef cfformat =
-        CTFontDescriptorCopyAttribute(fontd, kCTFontFormatAttribute);
-    if (!cfformat)
-        return false;
-    bool ret = is_postscript_font_format(cfformat);
-    CFRelease(cfformat);
-    return ret;
-}
-
 static bool check_glyph(void *priv, uint32_t code)
 {
     if (code == 0)
@@ -295,7 +272,6 @@ static void get_substitutions(void *priv, const char *name,
 }
 
 static ASS_FontProviderFuncs coretext_callbacks = {
-    .check_postscript   = check_postscript,
     .check_glyph        = check_glyph,
     .destroy_font       = destroy_font,
     .match_fonts        = match_fonts,

--- a/libass/ass_coretext.c
+++ b/libass/ass_coretext.c
@@ -57,8 +57,8 @@ static char *cfstr2buf(CFStringRef string)
 
 static void destroy_font(void *priv)
 {
-    CTFontDescriptorRef fontd = priv;
-    CFRelease(fontd);
+    CFCharacterSetRef set = priv;
+    SAFE_CFRelease(set);
 }
 
 static bool check_glyph(void *priv, uint32_t code)
@@ -66,16 +66,12 @@ static bool check_glyph(void *priv, uint32_t code)
     if (code == 0)
         return true;
 
-    CTFontDescriptorRef fontd = priv;
-    CFCharacterSetRef set =
-        CTFontDescriptorCopyAttribute(fontd, kCTFontCharacterSetAttribute);
+    CFCharacterSetRef set = priv;
 
     if (!set)
         return true;
 
-    bool result = CFCharacterSetIsLongCharacterMember(set, code);
-    CFRelease(set);
-    return result;
+    return CFCharacterSetIsLongCharacterMember(set, code);
 }
 
 static char *get_font_file(CTFontDescriptorRef fontd)
@@ -139,8 +135,9 @@ static void process_descriptors(ASS_FontProvider *provider, CFArrayRef fontsd)
 
         char *path = NULL;
         if (get_font_info_ct(fontd, &path, &meta)) {
-            CFRetain(fontd);
-            ass_font_provider_add_font(provider, &meta, path, index, (void*)fontd);
+            CFCharacterSetRef set =
+                CTFontDescriptorCopyAttribute(fontd, kCTFontCharacterSetAttribute);
+            ass_font_provider_add_font(provider, &meta, path, index, (void*)set);
         }
 
         free(meta.postscript_name);

--- a/libass/ass_coretext.c
+++ b/libass/ass_coretext.c
@@ -29,6 +29,18 @@
 
 #include "ass_coretext.h"
 
+#ifndef __has_builtin
+#define __has_builtin 0
+#endif
+
+#if __has_builtin(__builtin_available)
+#define CHECK_AVAILABLE(sym, ...) __builtin_available(__VA_ARGS__)
+#else
+// Cast to suppress "comparison never succeeds" warnings in some compilers
+// when the build target is new enough that sym isn't a weak symbol
+#define CHECK_AVAILABLE(sym, ...) (!!(intptr_t) &sym)
+#endif
+
 #define SAFE_CFRelease(x) do { if (x) CFRelease(x); } while (0)
 
 static const ASS_FontMapping font_substitutions[] = {

--- a/libass/ass_coretext.c
+++ b/libass/ass_coretext.c
@@ -19,6 +19,7 @@
 #include "config.h"
 #include "ass_compat.h"
 
+#include <Availability.h>
 #include <CoreFoundation/CoreFoundation.h>
 #include <TargetConditionals.h>
 #if TARGET_OS_IPHONE
@@ -88,7 +89,28 @@ static bool check_glyph(void *priv, uint32_t code)
 
 static char *get_font_file(CTFontDescriptorRef fontd)
 {
-    CFURLRef url = CTFontDescriptorCopyAttribute(fontd, kCTFontURLAttribute);
+    CFURLRef url = NULL;
+    if (false) {}
+#ifdef __MAC_10_6
+    // Declared in SDKs since 10.6, including iOS SDKs
+    else if (CHECK_AVAILABLE(kCTFontURLAttribute, macOS 10.6, *)) {
+        url = CTFontDescriptorCopyAttribute(fontd, kCTFontURLAttribute);
+    }
+#endif
+#if !TARGET_OS_IPHONE && (!defined(__MAC_10_6) || __MAC_OS_X_VERSION_MIN_REQUIRED < __MAC_10_6)
+    // ATS is declared deprecated in newer macOS SDKs
+    // and not declared at all in iOS SDKs
+    else {
+        CTFontRef font = CTFontCreateWithFontDescriptor(fontd, 0, NULL);
+        if (!font)
+            return NULL;
+        ATSFontRef ats_font = CTFontGetPlatformFont(font, NULL);
+        FSRef fs_ref;
+        if (ATSFontGetFileReference(ats_font, &fs_ref) == noErr)
+            url = CFURLCreateFromFSRef(NULL, &fs_ref);
+        CFRelease(font);
+    }
+#endif
     if (!url)
         return NULL;
     CFStringRef path = CFURLCopyFileSystemPath(url, kCFURLPOSIXPathStyle);

--- a/libass/ass_fontselect.c
+++ b/libass/ass_fontselect.c
@@ -78,6 +78,9 @@ struct font_info {
 
     // private data for callbacks
     void *priv;
+
+    // unused if the provider has a check_postscript function
+    bool is_postscript;
 };
 
 struct font_selector {
@@ -113,13 +116,6 @@ struct font_data_ft {
     FT_Face face;
     int idx;
 };
-
-static bool check_postscript_ft(void *data)
-{
-    FontDataFT *fd = (FontDataFT *)data;
-    PS_FontInfoRec postscript_info;
-    return !FT_Get_PS_Font_Info(fd->face, &postscript_info);
-}
 
 static bool check_glyph_ft(void *data, uint32_t codepoint)
 {
@@ -161,7 +157,6 @@ get_data_embedded(void *data, unsigned char *buf, size_t offset, size_t len)
 
 static ASS_FontProviderFuncs ft_funcs = {
     .get_data          = get_data_embedded,
-    .check_postscript  = check_postscript_ft,
     .check_glyph       = check_glyph_ft,
     .destroy_font      = destroy_font_ft,
 };
@@ -265,6 +260,7 @@ get_font_info(FT_Library lib, FT_Face face, const char *fallback_family_name,
     int slant, weight;
     char *fullnames[MAX_FULLNAME];
     char *families[MAX_FULLNAME];
+    PS_FontInfoRec postscript_info;
 
     // we're only interested in outlines
     if (!(face->face_flags & FT_FACE_FLAG_SCALABLE))
@@ -324,6 +320,7 @@ get_font_info(FT_Library lib, FT_Face face, const char *fallback_family_name,
     info->width  = 100;     // FIXME, should probably query the OS/2 table
 
     info->postscript_name = (char *)FT_Get_Postscript_Name(face);
+    info->is_postscript = !FT_Get_PS_Font_Info(face, &postscript_info);
 
     if (num_family) {
         info->families = calloc(sizeof(char *), num_family);
@@ -488,6 +485,7 @@ ass_font_provider_add_font(ASS_FontProvider *provider,
     info->width         = width;
     info->n_fullname    = meta->n_fullname;
     info->n_family      = meta->n_family;
+    info->is_postscript = meta->is_postscript;
 
     info->families = calloc(meta->n_family, sizeof(char *));
     if (info->families == NULL)
@@ -612,9 +610,12 @@ void ass_font_provider_free(ASS_FontProvider *provider)
 static bool check_postscript(ASS_FontInfo *fi)
 {
     ASS_FontProvider *provider = fi->provider;
-    assert(provider && provider->funcs.check_postscript);
+    assert(provider);
 
-    return provider->funcs.check_postscript(fi->priv);
+    if (provider->funcs.check_postscript)
+        return provider->funcs.check_postscript(fi->priv);
+    else
+        return fi->is_postscript;
 }
 
 /**

--- a/libass/ass_fontselect.h
+++ b/libass/ass_fontselect.h
@@ -158,7 +158,7 @@ typedef char   *(*GetFallbackFunc)(void *priv,
 
 typedef struct font_provider_funcs {
     GetDataFunc         get_data;               /* optional/mandatory */
-    CheckPostscriptFunc check_postscript;       /* mandatory */
+    CheckPostscriptFunc check_postscript;       /* optional */
     CheckGlyphFunc      check_glyph;            /* mandatory */
     DestroyFontFunc     destroy_font;           /* optional */
     DestroyProviderFunc destroy_provider;       /* optional */
@@ -211,6 +211,12 @@ struct ass_font_provider_meta_data {
                         // See FONT_WEIGHT_*
     int width;          // Font weight in percent, normally 100
                         // See FONT_WIDTH_*
+
+    /**
+     * Whether the font contains PostScript outlines.
+     * Unused if the font provider has a check_postscript function.
+     */
+    bool is_postscript;
 };
 
 struct ass_font_stream {

